File: projects/06-object-store/src/index_backend.rs
//! Index-as-a-service backend — From the field (ungraded).
//!
//! At S3 scale the key→location map is a **separately scaled metadata subsystem**
//! (persistence tier + coherent cache + witness), not a struct in the front-end
//! process. See [`RESEARCH.md` Part 2](../RESEARCH.md).
//!
//! - [`IndexBackend::Local`] delegates to the in-process [`Index`] (default when
//!   `INDEX_URL` is unset).
//! - [`IndexBackend::Remote`] / [`RemoteIndex`] speak HTTP JSON to the
//!   `object-store-index` binary when `INDEX_URL` is set.
//!
//! See `docs/05-how-index-as-a-service-works.md`.

use std::sync::Arc;

use serde::{Deserialize, Serialize};
use url::Url;

use crate::bucket::BucketMetadata;
use crate::error::AppError;
use crate::index::{Index, Listing, NewVersion, Precondition};
use crate::object::{Digest, ETag, ObjectMeta, ObjectRef, ResolvedObject, VersionId};

/// Body for `PUT /v1/{bucket}/keys/{key}` — publish a new live version.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PutRequest {
    pub digest: Digest,
    pub etag: ETag,
    pub size: u64,
    pub content_type: String,
    #[serde(default)]
    pub precondition: PreconditionWire,
}

impl PutRequest {
    pub fn into_parts(self) -> (NewVersion, Precondition) {
        (
            NewVersion {
                digest: self.digest,
                etag: self.etag,
                size: self.size,
                content_type: self.content_type,
            },
            self.precondition.into(),
        )
    }
}

/// Wire form of [`Precondition`] — kept separate so the on-disk index types
/// need not grow serde for HTTP.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum PreconditionWire {
    #[default]
    None,
    IfMatch {
        etag: ETag,
    },
    IfNoneMatchStar,
}

impl From<PreconditionWire> for Precondition {
    fn from(w: PreconditionWire) -> Self {
        match w {
            PreconditionWire::None => Precondition::None,
            PreconditionWire::IfMatch { etag } => Precondition::IfMatch(etag),
            PreconditionWire::IfNoneMatchStar => Precondition::IfNoneMatchStar,
        }
    }
}

impl From<Precondition> for PreconditionWire {
    fn from(p: Precondition) -> Self {
        match p {
            Precondition::None => PreconditionWire::None,
            Precondition::IfMatch(etag) => PreconditionWire::IfMatch { etag },
            Precondition::IfNoneMatchStar => PreconditionWire::IfNoneMatchStar,
        }
    }
}

/// Which version a resolve/delete addresses.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ObjectRefWire {
    Latest,
    Version { id: VersionId },
}

impl From<ObjectRefWire> for ObjectRef {
    fn from(w: ObjectRefWire) -> Self {
        match w {
            ObjectRefWire::Latest => ObjectRef::Latest,
            ObjectRefWire::Version { id } => ObjectRef::Version(id),
        }
    }
}

impl From<ObjectRef> for ObjectRefWire {
    fn from(r: ObjectRef) -> Self {
        match r {
            ObjectRef::Latest => ObjectRefWire::Latest,
            ObjectRef::Version(id) => ObjectRefWire::Version { id },
        }
    }
}

/// Query params for `GET /v1/{bucket}/list`.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ListQuery {
    #[serde(default)]
    pub prefix: String,
    pub delimiter: Option<String>,
    pub continuation: Option<String>,
    #[serde(default = "default_max_keys")]
    pub max_keys: usize,
}

fn default_max_keys() -> usize {
    1000
}

/// Wire form of [`Listing`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListingWire {
    pub objects: Vec<ObjectMeta>,
    pub common_prefixes: Vec<String>,
    pub next_continuation_token: Option<String>,
}

impl From<Listing> for ListingWire {
    fn from(l: Listing) -> Self {
        Self {
            objects: l.objects,
            common_prefixes: l.common_prefixes,
            next_continuation_token: l.next_continuation_token,
        }
    }
}

impl From<ListingWire> for Listing {
    fn from(l: ListingWire) -> Self {
        Self {
            objects: l.objects,
            common_prefixes: l.common_prefixes,
            next_continuation_token: l.next_continuation_token,
        }
    }
}

/// Wire form of [`ResolvedObject`] (domain type is not serde'd).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResolvedObjectWire {
    pub bucket: String,
    pub key: String,
    pub version_id: VersionId,
    pub digest: Digest,
    pub etag: ETag,
    pub size: u64,
    pub content_type: String,
    pub last_modified: chrono::DateTime<chrono::Utc>,
}

impl From<ResolvedObject> for ResolvedObjectWire {
    fn from(r: ResolvedObject) -> Self {
        Self {
            bucket: r.bucket,
            key: r.key,
            version_id: r.version_id,
            digest: r.digest,
            etag: r.etag,
            size: r.size,
            content_type: r.content_type,
            last_modified: r.last_modified,
        }
    }
}

impl From<ResolvedObjectWire> for ResolvedObject {
    fn from(r: ResolvedObjectWire) -> Self {
        Self {
            bucket: r.bucket,
            key: r.key,
            version_id: r.version_id,
            digest: r.digest,
            etag: r.etag,
            size: r.size,
            content_type: r.content_type,
            last_modified: r.last_modified,
        }
    }
}

/// Body for resolve / delete when the version is not in the path.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObjectRefBody {
    #[serde(default = "default_latest_ref")]
    pub object_ref: ObjectRefWire,
}

fn default_latest_ref() -> ObjectRefWire {
    ObjectRefWire::Latest
}

/// `POST /v1/gc` response.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GcResponse {
    pub reclaimed: u64,
}

/// JSON error body from the index service (mirrors S3 error codes by name).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndexErrorBody {
    pub code: String,
    pub message: String,
}

// ── Remote client ───────────────────────────────────────────────────────────

/// HTTP client for the index microservice (`object-store-index`).
#[derive(Debug, Clone)]
pub struct RemoteIndex {
    base_url: String,
    client: reqwest::Client,
}

impl RemoteIndex {
    pub fn new(base_url: impl Into<String>) -> Self {
        Self {
            base_url: base_url.into().trim_end_matches('/').to_string(),
            client: reqwest::Client::new(),
        }
    }

    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    fn parse_base(&self) -> Result<Url, AppError> {
        Url::parse(&self.base_url)
            .map_err(|e| AppError::InvalidRequest(format!("bad INDEX_URL: {e}")))
    }

    fn with_segments(&self, segments: &[&str]) -> Result<Url, AppError> {
        let mut url = self.parse_base()?;
        {
            let mut path = url.path_segments_mut().map_err(|_| {
                AppError::InvalidRequest("INDEX_URL cannot be a base for path joins".into())
            })?;
            // Drop an empty trailing segment from `http://host:port/`.
            path.clear();
            for s in segments {
                path.push(s);
            }
        }
        Ok(url)
    }

    fn bucket_url(&self, bucket: &str) -> Result<Url, AppError> {
        self.with_segments(&["v1", "buckets", bucket])
    }

    fn key_url(&self, bucket: &str, key: &str) -> Result<Url, AppError> {
        let mut url = self.parse_base()?;
        {
            let mut path = url.path_segments_mut().map_err(|_| {
                AppError::InvalidRequest("INDEX_URL cannot be a base for path joins".into())
            })?;
            path.clear();
            path.extend(["v1", "buckets", bucket, "keys"]);
            for segment in key.split('/') {
                path.push(segment);
            }
        }
        Ok(url)
    }

    async fn map_response_error(&self, res: reqwest::Response) -> AppError {
        let status = res.status();
        if let Ok(body) = res.json::<IndexErrorBody>().await {
            return app_error_from_wire(&body.code, body.message);
        }
        AppError::Other(anyhow::anyhow!(
            "index service HTTP {status} (unparseable error body)"
        ))
    }

    /// Send `req`, map transport errors, and require a success status (no body).
    async fn request_empty(&self, req: reqwest::RequestBuilder) -> Result<(), AppError> {
        let res = req.send().await.map_err(|e| AppError::Other(e.into()))?;
        if res.status().is_success() {
            return Ok(());
        }
        Err(self.map_response_error(res).await)
    }

    /// Send `req`, map transport errors, and decode a JSON success body.
    async fn request_json<T: for<'de> Deserialize<'de>>(
        &self,
        req: reqwest::RequestBuilder,
    ) -> Result<T, AppError> {
        let res = req.send().await.map_err(|e| AppError::Other(e.into()))?;
        if res.status().is_success() {
            return res
                .json()
                .await
                .map_err(|e| AppError::Other(anyhow::anyhow!("index JSON decode: {e}")));
        }
        Err(self.map_response_error(res).await)
    }

    pub async fn create_bucket(&self, bucket: &str) -> Result<(), AppError> {
        self.request_empty(self.client.put(self.bucket_url(bucket)?))
            .await
    }

    pub async fn buckets(&self) -> Result<Vec<String>, AppError> {
        let url = self.with_segments(&["v1", "buckets"])?;
        self.request_json(self.client.get(url)).await
    }

    pub async fn ensure_bucket(&self, bucket: &str) -> Result<(), AppError> {
        self.request_empty(self.client.head(self.bucket_url(bucket)?))
            .await
    }

    pub async fn put(
        &self,
        bucket: &str,
        key: &str,
        version: NewVersion,
        pre: Precondition,
    ) -> Result<ObjectMeta, AppError> {
        let url = self.key_url(bucket, key)?;
        let body = PutRequest {
            digest: version.digest,
            etag: version.etag,
            size: version.size,
            content_type: version.content_type,
            precondition: pre.into(),
        };
        self.request_json(self.client.put(url).json(&body)).await
    }

    pub async fn get(&self, bucket: &str, key: &str) -> Result<Option<ObjectMeta>, AppError> {
        let url = self.key_url(bucket, key)?;
        self.request_json(self.client.get(url)).await
    }

    pub async fn resolve(
        &self,
        bucket: &str,
        key: &str,
        object_ref: ObjectRef,
    ) -> Result<ResolvedObject, AppError> {
        // Separate prefix from `/keys/{*key}` — axum cannot nest under a wildcard.
        let mut url = self.parse_base()?;
        {
            let mut path = url.path_segments_mut().map_err(|_| {
                AppError::InvalidRequest("INDEX_URL cannot be a base for path joins".into())
            })?;
            path.clear();
            path.extend(["v1", "buckets", bucket, "resolve"]);
            for segment in key.split('/') {
                path.push(segment);
            }
        }
        let body = ObjectRefBody {
            object_ref: object_ref.into(),
        };
        let wire: ResolvedObjectWire = self.request_json(self.client.post(url).json(&body)).await?;
        Ok(wire.into())
    }

    pub async fn delete(
        &self,
        bucket: &str,
        key: &str,
        object_ref: ObjectRef,
    ) -> Result<(), AppError> {
        let url = self.key_url(bucket, key)?;
        let body = ObjectRefBody {
            object_ref: object_ref.into(),
        };
        self.request_empty(self.client.delete(url).json(&body))
            .await
    }

    pub async fn list(
        &self,
        bucket: &str,
        prefix: &str,
        delimiter: Option<&str>,
        continuation: Option<&str>,
        max_keys: usize,
    ) -> Result<Listing, AppError> {
        let mut url = self.with_segments(&["v1", "buckets", bucket, "list"])?;
        {
            let mut q = url.query_pairs_mut();
            q.append_pair("prefix", prefix);
            q.append_pair("max_keys", &max_keys.to_string());
            if let Some(d) = delimiter {
                q.append_pair("delimiter", d);
            }
            if let Some(c) = continuation {
                q.append_pair("continuation", c);
            }
        }
        let wire: ListingWire = self.request_json(self.client.get(url)).await?;
        Ok(wire.into())
    }

    pub async fn index_entries(&self, bucket: &str) -> Result<Vec<ObjectMeta>, AppError> {
        let url = self.with_segments(&["v1", "buckets", bucket, "entries"])?;
        self.request_json(self.client.get(url)).await
    }

    pub async fn gc(&self) -> Result<u64, AppError> {
        let url = self.with_segments(&["v1", "gc"])?;
        let body: GcResponse = self.request_json(self.client.post(url)).await?;
        Ok(body.reclaimed)
    }

    pub async fn load_bucket_metadata(&self, bucket: &str) -> Result<BucketMetadata, AppError> {
        let url = self.with_segments(&["v1", "buckets", bucket, "metadata"])?;
        self.request_json(self.client.get(url)).await
    }

    pub async fn store_bucket_metadata(
        &self,
        bucket: &str,
        meta: &BucketMetadata,
    ) -> Result<(), AppError> {
        let url = self.with_segments(&["v1", "buckets", bucket, "metadata"])?;
        self.request_empty(self.client.put(url).json(meta)).await
    }
}

fn app_error_from_wire(code: &str, message: String) -> AppError {
    match code {
        "NoSuchBucket" => AppError::NoSuchBucket,
        "NoSuchKey" => AppError::NoSuchKey,
        "NoSuchUpload" => AppError::NoSuchUpload,
        "BucketAlreadyExists" => AppError::BucketAlreadyExists,
        "InvalidRequest" => AppError::InvalidRequest(message),
        "EntityTooLarge" => AppError::EntityTooLarge,
        "PreconditionFailed" => AppError::PreconditionFailed,
        _ => AppError::Other(anyhow::anyhow!("index service: {code}: {message}")),
    }
}

/// Process-local or RPC-backed index. [`AppState`] holds `Arc<IndexBackend>`;
/// set `INDEX_URL` to pick [`Self::Remote`].
#[derive(Clone)]
pub enum IndexBackend {
    Local(Arc<Index>),
    Remote(RemoteIndex),
}

impl IndexBackend {
    pub fn local(index: Arc<Index>) -> Self {
        Self::Local(index)
    }

    pub fn remote(base_url: impl Into<String>) -> Self {
        Self::Remote(RemoteIndex::new(base_url))
    }

    pub async fn create_bucket(&self, bucket: &str) -> Result<(), AppError> {
        match self {
            Self::Local(i) => i.create_bucket(bucket).await,
            Self::Remote(r) => r.create_bucket(bucket).await,
        }
    }

    pub async fn buckets(&self) -> Result<Vec<String>, AppError> {
        match self {
            Self::Local(i) => i.buckets().await,
            Self::Remote(r) => r.buckets().await,
        }
    }

    /// Network-friendly ensure: existence check only (no [`std::path::PathBuf`]).
    pub async fn ensure_bucket(&self, bucket: &str) -> Result<(), AppError> {
        match self {
            Self::Local(i) => {
                i.ensure_bucket(bucket).await?;
                Ok(())
            }
            Self::Remote(r) => r.ensure_bucket(bucket).await,
        }
    }

    pub async fn put(
        &self,
        bucket: &str,
        key: &str,
        version: NewVersion,
        pre: Precondition,
    ) -> Result<ObjectMeta, AppError> {
        match self {
            Self::Local(i) => i.put(bucket, key, version, pre).await,
            Self::Remote(r) => r.put(bucket, key, version, pre).await,
        }
    }

    pub async fn get(&self, bucket: &str, key: &str) -> Result<Option<ObjectMeta>, AppError> {
        match self {
            Self::Local(i) => i.get(bucket, key).await,
            Self::Remote(r) => r.get(bucket, key).await,
        }
    }

    pub async fn resolve(
        &self,
        bucket: &str,
        key: &str,
        object_ref: ObjectRef,
    ) -> Result<ResolvedObject, AppError> {
        match self {
            Self::Local(i) => i.resolve(bucket, key, object_ref).await,
            Self::Remote(r) => r.resolve(bucket, key, object_ref).await,
        }
    }

    pub async fn delete(
        &self,
        bucket: &str,
        key: &str,
        object_ref: ObjectRef,
    ) -> Result<(), AppError> {
        match self {
            Self::Local(i) => i.delete(bucket, key, object_ref).await,
            Self::Remote(r) => r.delete(bucket, key, object_ref).await,
        }
    }

    pub async fn list(
        &self,
        bucket: &str,
        prefix: &str,
        delimiter: Option<&str>,
        continuation: Option<&str>,
        max_keys: usize,
    ) -> Result<Listing, AppError> {
        match self {
            Self::Local(i) => {
                i.list(bucket, prefix, delimiter, continuation, max_keys)
                    .await
            }
            Self::Remote(r) => {
                r.list(bucket, prefix, delimiter, continuation, max_keys)
                    .await
            }
        }
    }

    pub async fn index_entries(&self, bucket: &str) -> Result<Vec<ObjectMeta>, AppError> {
        match self {
            Self::Local(i) => i.index_entries(bucket).await,
            Self::Remote(r) => r.index_entries(bucket).await,
        }
    }

    pub async fn gc(&self) -> Result<u64, AppError> {
        match self {
            Self::Local(i) => i.gc().await,
            Self::Remote(r) => r.gc().await,
        }
    }

    /// Load `index/<bucket>/metadata.json` (locally or via the index service).
    pub async fn load_bucket_metadata(&self, bucket: &str) -> Result<BucketMetadata, AppError> {
        match self {
            Self::Local(i) => {
                let dir = i.ensure_bucket(bucket).await?;
                BucketMetadata::load(&dir).await
            }
            Self::Remote(r) => r.load_bucket_metadata(bucket).await,
        }
    }

    /// Persist bucket metadata (lifecycle policy, etc.).
    pub async fn store_bucket_metadata(
        &self,
        bucket: &str,
        meta: &BucketMetadata,
    ) -> Result<(), AppError> {
        match self {
            Self::Local(i) => {
                let dir = i.ensure_bucket(bucket).await?;
                meta.store(&dir).await
            }
            Self::Remote(r) => r.store_bucket_metadata(bucket, meta).await,
        }
    }
}

File: projects/06-object-store/src/object_store/index_backend.py
"""Index-as-a-service — From the field, ungraded.

At S3 scale the key→location map is a **separately scaled metadata subsystem**,
not a struct inside the front-end process. Metadata operations and data
operations have completely different shapes: a PUT of a 5 GB object is one index
write and 5 GB of disk I/O, while a listing is thousands of index reads and no
data at all. Scaling them together means over-provisioning one to serve the
other.

Splitting them also makes the blob-then-pointer invariant *visible*: with the
index on the far side of a network call, "the blob is durable before the pointer
exists" stops being an ordering you can accidentally reverse in one function and
becomes an ordering between two systems. Kill the index process mid-PUT and you
get exactly the failure the invariant predicts — an orphan blob, never a
dangling key.

- `LocalIndex` delegates to the in-process `Index` (the default).
- `RemoteIndex` speaks HTTP JSON to the `object-store-index` process when
  `INDEX_URL` is set. Blobs still live under the front-end's own data dir.

Both satisfy `IndexBackend`, so nothing upstream knows which one it has.

## Errors have to survive the wire

The interesting part of the remote client is not the happy path — it is that a
`NoSuchKey` raised inside the index process must arrive at the S3 handler as a
`NoSuchKey`, not as "HTTP 404". So the service sends `{"code", "message"}` and
this module maps the code back onto the exception class. Lose that and every
error in the split deployment degrades to a generic 500, and the conditional
writes (which depend on 412 specifically) stop working.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Protocol, runtime_checkable

import httpx
import structlog

from .bucket import BucketMetadata
from .errors import (
    AccessDenied,
    AppError,
    BucketAlreadyExists,
    EntityTooLarge,
    InvalidRequest,
    NoSuchBucket,
    NoSuchKey,
    NoSuchUpload,
    PreconditionFailed,
)
from .index import Index, Listing, NewVersion, Precondition
from .naming import Bucket, Key
from .objects import LATEST, BlobKind, Digest, ETag, ObjectMeta, ObjectRef, ResolvedObject

if TYPE_CHECKING:
    from .config import Settings
    from .store import Store

__all__ = ["IndexBackend", "LocalIndex", "RemoteIndex", "make_index_backend"]

logger = structlog.get_logger(__name__)

_ERROR_CODES: dict[str, type[AppError]] = {
    "NoSuchBucket": NoSuchBucket,
    "NoSuchKey": NoSuchKey,
    "NoSuchUpload": NoSuchUpload,
    "BucketAlreadyExists": BucketAlreadyExists,
    "InvalidRequest": InvalidRequest,
    "EntityTooLarge": EntityTooLarge,
    "PreconditionFailed": PreconditionFailed,
    "AccessDenied": AccessDenied,
}


@runtime_checkable
class IndexBackend(Protocol):
    """What the S3 layer needs from an index, wherever it lives."""

    async def create_bucket(self, bucket: Bucket) -> None: ...
    async def buckets(self) -> list[str]: ...
    async def ensure_bucket(self, bucket: Bucket) -> None: ...
    async def put(
        self,
        bucket: Bucket,
        key: Key,
        version: NewVersion,
        precondition: Precondition | None = None,
    ) -> ObjectMeta: ...
    async def get(self, bucket: Bucket, key: Key) -> ObjectMeta | None: ...
    async def resolve(
        self, bucket: Bucket, key: Key, object_ref: ObjectRef = LATEST
    ) -> ResolvedObject: ...
    async def delete(self, bucket: Bucket, key: Key, object_ref: ObjectRef = LATEST) -> None: ...
    async def list(
        self,
        bucket: Bucket,
        prefix: str = "",
        delimiter: str | None = None,
        continuation: str | None = None,
        max_keys: int = 1000,
    ) -> Listing: ...
    async def index_entries(self, bucket: Bucket) -> list[ObjectMeta]: ...
    async def gc(self) -> int: ...
    async def load_bucket_metadata(self, bucket: Bucket) -> BucketMetadata: ...
    async def store_bucket_metadata(self, bucket: Bucket, metadata: BucketMetadata) -> None: ...
    async def aclose(self) -> None: ...


class LocalIndex:
    """The in-process index, wrapped to satisfy `IndexBackend`.

    Almost all delegation, with one real difference: `ensure_bucket` returns a
    `Path` on `Index` because callers there want the directory, and returns
    nothing here because a path is meaningless across a network. Narrowing it is
    what lets the remote backend implement the same protocol at all.
    """

    __slots__ = ("index",)

    def __init__(self, index: Index) -> None:
        self.index = index

    async def create_bucket(self, bucket: Bucket) -> None:
        await self.index.create_bucket(bucket)

    async def buckets(self) -> list[str]:
        return await self.index.buckets()

    async def ensure_bucket(self, bucket: Bucket) -> None:
        await self.index.ensure_bucket(bucket)

    async def put(
        self,
        bucket: Bucket,
        key: Key,
        version: NewVersion,
        precondition: Precondition | None = None,
    ) -> ObjectMeta:
        return await self.index.put(bucket, key, version, precondition)

    async def get(self, bucket: Bucket, key: Key) -> ObjectMeta | None:
        return await self.index.get(bucket, key)

    async def resolve(
        self, bucket: Bucket, key: Key, object_ref: ObjectRef = LATEST
    ) -> ResolvedObject:
        return await self.index.resolve(bucket, key, object_ref)

    async def delete(self, bucket: Bucket, key: Key, object_ref: ObjectRef = LATEST) -> None:
        await self.index.delete(bucket, key, object_ref)

    async def list(
        self,
        bucket: Bucket,
        prefix: str = "",
        delimiter: str | None = None,
        continuation: str | None = None,
        max_keys: int = 1000,
    ) -> Listing:
        return await self.index.list(bucket, prefix, delimiter, continuation, max_keys)

    async def index_entries(self, bucket: Bucket) -> list[ObjectMeta]:
        return await self.index.index_entries(bucket)

    async def gc(self) -> int:
        return await self.index.gc()

    async def load_bucket_metadata(self, bucket: Bucket) -> BucketMetadata:
        return await self.index.load_bucket_metadata(bucket)

    async def store_bucket_metadata(self, bucket: Bucket, metadata: BucketMetadata) -> None:
        await self.index.store_bucket_metadata(bucket, metadata)

    async def aclose(self) -> None:
        """Nothing to close — there is no connection."""


def _precondition_wire(precondition: Precondition | None) -> dict[str, Any]:
    """Serialise a precondition for the wire.

    A separate shape from `Precondition` itself so the in-process type never has
    to grow serialisation concerns it does not otherwise need.
    """
    guard = precondition or Precondition.none()
    if guard.kind == "if_match":
        return {"type": "if_match", "etag": str(guard.etag)}
    if guard.kind == "if_none_match_star":
        return {"type": "if_none_match_star"}
    return {"type": "none"}


def precondition_from_wire(payload: dict[str, Any]) -> Precondition:
    kind = payload.get("type", "none")
    if kind == "if_match":
        return Precondition.if_match(ETag(payload["etag"]))
    if kind == "if_none_match_star":
        return Precondition.if_none_match_star()
    return Precondition.none()


def _object_ref_wire(object_ref: ObjectRef) -> dict[str, Any]:
    if object_ref.is_latest:
        return {"type": "latest"}
    return {"type": "version", "id": object_ref.version_id}


def object_ref_from_wire(payload: dict[str, Any]) -> ObjectRef:
    if payload.get("type") == "version":
        return ObjectRef.version(int(payload["id"]))
    return LATEST


class RemoteIndex:
    """HTTP client for the `object-store-index` process."""

    __slots__ = ("_client", "base_url")

    def __init__(self, base_url: str, client: httpx.AsyncClient | None = None) -> None:
        self.base_url = base_url.rstrip("/")
        # One client, reused: it holds the connection pool, and building a fresh
        # one per call would pay a TCP and TLS handshake on every index read —
        # which on the GET path is once per request.
        self._client = client or httpx.AsyncClient(timeout=10.0)

    async def aclose(self) -> None:
        await self._client.aclose()

    def _url(self, *segments: str) -> str:
        from urllib.parse import quote

        path = "/".join(quote(segment, safe="") for segment in segments)
        return f"{self.base_url}/{path}"

    def _key_url(self, bucket: Bucket, key: Key, *tail: str) -> str:
        # `safe="/"` keeps the key's slashes as path separators, matching the
        # service's `{key:path}` route. Fully escaping them does *not* work:
        # ASGI servers normalise `%2F` back to a separator before routing, so
        # the encoded form never reaches a single-segment route at all. Every
        # other character is still escaped, so a key containing a literal `%2f`
        # is sent as `%252f` and stays distinct from one containing `/`.
        from urllib.parse import quote

        prefix = self._url("v1", "buckets", bucket, *tail)
        return f"{prefix}/{quote(key, safe='/')}"

    async def _request(self, method: str, url: str, **kwargs: Any) -> Any:
        try:
            response = await self._client.request(method, url, **kwargs)
        except httpx.HTTPError as err:
            raise AppError(f"index service unreachable: {err}") from err

        if response.is_success:
            if not response.content:
                return None
            return response.json()

        raise _error_from_response(response)

    async def create_bucket(self, bucket: Bucket) -> None:
        await self._request("PUT", self._url("v1", "buckets", bucket))

    async def buckets(self) -> list[str]:
        return await self._request("GET", self._url("v1", "buckets"))

    async def ensure_bucket(self, bucket: Bucket) -> None:
        # GET, not HEAD — see the service's route for why the body matters.
        await self._request("GET", self._url("v1", "buckets", bucket))

    async def put(
        self,
        bucket: Bucket,
        key: Key,
        version: NewVersion,
        precondition: Precondition | None = None,
    ) -> ObjectMeta:
        payload = await self._request(
            "PUT",
            self._key_url(bucket, key, "keys"),
            json={
                "digest": str(version.digest),
                "etag": str(version.etag),
                "size": version.size,
                "content_type": version.content_type,
                "blob_kind": version.blob_kind.value,
                "precondition": _precondition_wire(precondition),
            },
        )
        return ObjectMeta.model_validate(payload)

    async def get(self, bucket: Bucket, key: Key) -> ObjectMeta | None:
        payload = await self._request("GET", self._key_url(bucket, key, "keys"))
        return ObjectMeta.model_validate(payload) if payload else None

    async def resolve(
        self, bucket: Bucket, key: Key, object_ref: ObjectRef = LATEST
    ) -> ResolvedObject:
        payload = await self._request(
            "POST",
            self._key_url(bucket, key, "resolve"),
            json={"object_ref": _object_ref_wire(object_ref)},
        )
        return ResolvedObject(
            bucket=Bucket.from_trusted(payload["bucket"]),
            key=Key.from_trusted(payload["key"]),
            version_id=payload["version_id"],
            digest=Digest(payload["digest"]),
            etag=ETag(payload["etag"]),
            size=payload["size"],
            content_type=payload["content_type"],
            last_modified=_parse_datetime(payload["last_modified"]),
            blob_kind=BlobKind(payload.get("blob_kind", "whole")),
        )

    async def delete(self, bucket: Bucket, key: Key, object_ref: ObjectRef = LATEST) -> None:
        await self._request(
            "DELETE",
            self._key_url(bucket, key, "keys"),
            json={"object_ref": _object_ref_wire(object_ref)},
        )

    async def list(
        self,
        bucket: Bucket,
        prefix: str = "",
        delimiter: str | None = None,
        continuation: str | None = None,
        max_keys: int = 1000,
    ) -> Listing:
        params: dict[str, str] = {"prefix": prefix, "max_keys": str(max_keys)}
        if delimiter:
            params["delimiter"] = delimiter
        if continuation:
            params["continuation"] = continuation
        payload = await self._request(
            "GET", self._url("v1", "buckets", bucket, "list"), params=params
        )
        return Listing(
            objects=[ObjectMeta.model_validate(row) for row in payload["objects"]],
            common_prefixes=payload["common_prefixes"],
            next_continuation_token=payload.get("next_continuation_token"),
        )

    async def index_entries(self, bucket: Bucket) -> list[ObjectMeta]:
        payload = await self._request("GET", self._url("v1", "buckets", bucket, "entries"))
        return [ObjectMeta.model_validate(row) for row in payload]

    async def gc(self) -> int:
        payload = await self._request("POST", self._url("v1", "gc"))
        return int(payload["reclaimed"])

    async def load_bucket_metadata(self, bucket: Bucket) -> BucketMetadata:
        payload = await self._request("GET", self._url("v1", "buckets", bucket, "metadata"))
        return BucketMetadata.model_validate(payload)

    async def store_bucket_metadata(self, bucket: Bucket, metadata: BucketMetadata) -> None:
        await self._request(
            "PUT",
            self._url("v1", "buckets", bucket, "metadata"),
            content=metadata.model_dump_json().encode("utf-8"),
            headers={"content-type": "application/json"},
        )


def _error_from_response(response: httpx.Response) -> AppError:
    """Rebuild the original exception from the service's JSON error body."""
    try:
        payload = response.json()
        code = str(payload["code"])
        message = str(payload.get("message", ""))
    except (ValueError, KeyError, TypeError):
        fallback = _STATUS_FALLBACK.get(response.status_code)
        if fallback is not None:
            return fallback()
        return AppError(f"index service HTTP {response.status_code} (unparseable error body)")
    factory = _ERROR_CODES.get(code)
    if factory is None:
        return AppError(f"index service: {code}: {message}")
    return factory(message) if message else factory()


_STATUS_FALLBACK: dict[int, type[AppError]] = {
    404: NoSuchKey,
    409: BucketAlreadyExists,
    400: InvalidRequest,
    412: PreconditionFailed,
    413: EntityTooLarge,
    403: AccessDenied,
}
"""Used only when a response has no body to read the code out of — a proxy that
stripped one, or a error the service did not shape.

Deliberately a last resort: the status code is coarser than the S3 code (404
alone covers `NoSuchBucket`, `NoSuchKey` and `NoSuchUpload`), so every endpoint
here is designed to answer with a body rather than lean on this."""


def _parse_datetime(raw: str):  # noqa: ANN202
    from datetime import datetime

    return datetime.fromisoformat(raw)


def make_index_backend(settings: Settings, store: Store) -> IndexBackend:
    """Pick a backend from configuration.

    An empty `INDEX_URL` means in-process, which is the default and what tests
    use. Setting it is the only thing needed to move metadata to its own
    process — the S3 layer's code does not change.
    """
    if settings.index_url:
        logger.info("using the remote index service", index_url=settings.index_url)
        return RemoteIndex(settings.index_url)
    return LocalIndex(Index(settings.data_dir, store))
